xarray/core/dataarray.py

```diff
@@ -1736,5 +1736,93 @@ def dot(self, other):
 
         return type(self)(new_data, new_coords, new_dims)
 
+    def quantile(self, q, dim=None, axis=None, interpolation='linear'):
```

Comment:
Probably the most consistent with the current approach is to implement the guts (calling `np.nanpercentile`) on `Variable`, and then make a wrapper for handling coordinates on `Dataset` (and call that from the `DataArray` method).